Support partial Process Stats responses when the stats service (log-cache or metric-proxy) is unavailable

tcdowney opened the pull request:
This change rescues errors from log-cache and metric-proxy and passes them up as warnings to callers instead of exceptions so that the stats endpoint can continue to function on a best-effort basis. On /v3/processes/:guid/stats these warnings are included in a new "warnings" key on the response, on v2 they are ignored. Alternatively/additionally we could add these to the `'X-Cf-Warnings'` header as @sethboyles [mentioned on the issue](https://github.com/cloudfoundry/cloud_controller_ng/issues/2227#issuecomment-828039303).

The related issue has more information: https://github.com/cloudfoundry/cloud_controller_ng/issues/2227

* [x] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have run all the unit tests using `bundle exec rake`

* [x] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats)

sethboyles commented on lib/cloud_controller/diego/reporters/instances_stats_reporter.rb:
```diff
@@ -92,15 +121,16 @@ def envelopes(desired_lrp, process)
-        )
+        ), nil
+      rescue GRPC::BadStatus, CloudController::Errors::ApiError => e
+        logger.error('stats_for_app.error', error: e.message, backtrace: e.backtrace.join($INPUT_RECORD_SEPARATOR))
```
what does this global variable do?

tcdowney replied:
It just joins them with the system's concept of what the newline separator should be. I think it's always `"\n"` in our case and it's probably better if I just explicitly declare that.

tcdowney replied:
Made the change:
https://github.com/cloudfoundry/cloud_controller_ng/pull/2250/files#diff-ad5dff07755b8d42aeadbf674ff7947d38ac428c9e7e52e415a6194d07ba202eR130

sethboyles commented:
so far looks good to me.  Is the non-idiomatic part of this you mentioned the multiple return values?  If so I don't have any qualms with it.

tcdowney commented:
> so far looks good to me. Is the non-idiomatic part of this you mentioned the multiple return values? If so I don't have any qualms with it.

Yeah, that was mostly it. I don't think we've done much of it in the codebase so wanted opinions. I think it's better than trying to juggle the exceptions all the way up to the controller in this case, though.

tcdowney commented:
@sethboyles do you think we should go with the warning header instead of the `"warnings"` field in the response? I could see that being a better UX out of the box since the CLI already supports it.

sethboyles commented:
I think for now that might be the better option.  I am uncertain about the `"warnings"` property as it has only been used for jobs in the past, and I am wary of introducing it to other parts of the API without giving it enough thought or discussion. 
 Clients probably don't have existing ways of consuming non-job related warnings, and I'm unsure its a pattern we want to adopt generally as we probably don't want to allow partial successes/failures when we can avoid it.  

We are going to discuss warnings in general as a team tomorrow, but I don't want to block this PR--how about we go with the header warning for now, and if we decide later on to use the JSON field, we can put it in our backlog?

tcdowney commented:
Sounds good. I'll make that change. 👍

tcdowney commented:
## Validating Steps
Here's how I tested this out manually.

### CF for VMs
1.  `cf push dora` and observe the metrics work
2. `cf restart dora` and observe the metrics work
3. `bosh ssh doppler`
4. `monit stop log-cache`
5. `cf restart dora` and observe that the workflow succeeds even though metrics are missing. My output looked like this:

```console
$ cf restart dora
Restarting app dora in org org / space space as admin...

Stopping app...

Waiting for app to start...

Instances starting...
Instances starting...
Instances starting...
Instances starting...

Stats server temporarily unavailable.
Stats server temporarily unavailable.
Stats server temporarily unavailable.
Stats server temporarily unavailable.
Stats server temporarily unavailable.
No running instances found for process guid 63461f63-0ab9-45c0-bc16-482af9179d3c-7219907a-a412-45cd-a4f8-8fe6bd6477f6
name:              dora
requested state:   started
routes:            dora.diamond-legend.capi.land
last uploaded:     Thu 29 Apr 09:20:20 PDT 2021
stack:             cflinuxfs3
buildpacks:
	name             version   detect output   buildpack name
	ruby_buildpack   1.8.37    ruby            ruby

type:           web
sidecars:
instances:      1/1
memory usage:   256M
     state     since                  cpu    memory   disk     details
#0   running   2021-04-29T16:22:54Z   0.0%   0 of 0   0 of 0

type:           worker
sidecars:
instances:      0/0
memory usage:   256M
There are no running instances of this process.
```

Directly hitting the endpoint returns:
```console
cf curl /v3/processes/8f79dd94-2f93-427c-b66c-9290b81db79d/stats
{
   "resources": [
      {
         "type": "web",
         "index": 0,
         "state": "RUNNING",
         "host": "10.244.0.138",
         "uptime": 258,
         "mem_quota": null,
         "disk_quota": null,
         "fds_quota": 16384,
         "isolation_segment": null,
         "details": null,
         "instance_ports": [
            {
               "external": 61004,
               "internal": 8080,
               "external_tls_proxy_port": 61006,
               "internal_tls_proxy_port": 61001
            },
            {
               "external": 61005,
               "internal": 2222,
               "external_tls_proxy_port": 61007,
               "internal_tls_proxy_port": 61002
            }
         ],
         "usage": {}
      }
   ]
}
```

`cf curl` also prints out the `Stats server temporarily unavailable.` warning after it finishes.

Even though the API is returning `null`/empty values,  CLI is reporting 0s for the metrics that is the default numeric value in Go. Since we're returning empty values on the API side, the CLI could be enhanced to display something more meaningful if it wanted to, but I think this is fine for now.

### CF for Kubernetes
On Kubernetes the CLI/API experience is the same, just the repro steps are a little different.
1.  `cf push dora` and observe the metrics work (on cf-for-k8s I always see the quotas missing even without this change)
2. `cf restart dora` and observe the metrics work (on cf-for-k8s I always see the quotas missing even without this change)
3. `kubectl -n cf-system scale deployment metric-proxy --replicas=0`
4. `cf restart dora` and observe that the workflow succeeds even though metrics are missing.

tcdowney commented:
@sethboyles I rebased the original commit to use the `'X-Cf-Warnings'` header instead of adding `"warnings"` to the response. Also added a commit to remove a now unused config flag and some light-weight documentation to point out these fields can now be empty or `null` (the `mem_quota` and `disk_quota` ones could have been `null` before this change fwiw).

I'm gonna take this PR out of "draft" state now, I think it's ready. 🙂

jspawar commented on app/controllers/runtime/stats_controller.rb:
```diff
@@ -22,7 +22,7 @@ def stats(guid, opts={})
 
       begin
-        stats = instances_reporters.stats_for_app(process)
+        stats, _ = instances_reporters.stats_for_app(process)
```
Why do we ignore warnings in v2?

tcdowney replied:
> I don't quite understand why we ignore the errors in v2, although I'm guessing this is because v2 endpoints should not be modified and they don't support that warnings header?

Yeah, I didn't want to change v2 too much. It does have code in place that would support adding the header, though: https://github.com/cloudfoundry/cloud_controller_ng/blob/4a23a419862ef9b7272f411a4435e33fc7305f30/app/controllers/base/base_controller.rb#L124

sethboyles replied:
Sounds good to me!

tcdowney replied:
932e8ecac467ae656432a38f49a68b240c380a74

jspawar reviewed (approved):
Looks good, I don't quite understand why we ignore the errors in v2, although I'm guessing this is because v2 endpoints should not be modified and they don't support that warnings header?

sethboyles commented on docs/v3/source/includes/resources/processes/_stats_object.md.erb:
```diff
@@ -12,15 +12,16 @@ Name | Type | Description
-**mem_quota** | _integer_ | The current maximum memory allocated for the instance
-**disk_quota** | _integer_ | The current maximum disk allocated for the instance
+**mem_quota** | _integer_ | The current maximum memory allocated for the instance; the value is `null` when memory quota data is unavailable
+**disk_quota** | _integer_ | The current maximum disk allocated for the instance; the value is `null` when disk quota data is unavailable
```
🙏  thank you for updating the docs!

sethboyles commented:
LGTM!

```
± sb |handle-log-cache-unavailability → origin {2} ?:1 ✗| → cf curl v3/apps/$(cf app dora --guid)/processes/web/stats
{
   "resources": [
      {
         "type": "web",
         "index": 0,
         "state": "RUNNING",
         "host": "10.244.0.138",
         "uptime": 1525,
         "mem_quota": null,
         "disk_quota": null,
         "fds_quota": 16384,
         "isolation_segment": null,
         "details": null,
         "instance_ports": [
            {
               "external": 61028,
               "internal": 8080,
               "external_tls_proxy_port": 61030,
               "internal_tls_proxy_port": 61001
            },
            {
               "external": 61029,
               "internal": 2222,
               "external_tls_proxy_port": 61031,
               "internal_tls_proxy_port": 61002
            }
         ],
         "usage": {}
      }
   ]
}
Stats server temporarily unavailable.
```

@tcdowney Let me know if/when you add the warnings to v2, then I'll go ahead and merge this in.  Thank you for the PR!

tcdowney commented:
```console
19:04 $ cf curl /v2/apps/$(cf app dora --guid)/stats
{
   "0": {
      "state": "RUNNING",
      "isolation_segment": null,
      "stats": {
         "name": "dora",
         "uris": [
            "dora.diamond-legend.capi.land"
         ],
         "host": "10.244.0.138",
         "port": 61004,
         "uptime": 62,
         "fds_quota": 16384,
         "mem_quota": null,
         "disk_quota": null,
         "usage": {}
      }
   }
}
Stats server temporarily unavailable.
```

Done! 👍
